### PR TITLE
kubeadm: Migrate to client-go

### PR DESCRIPTION
**Pull request by dmmcquay**
**What this PR does / why we need it**: Finish the migration for kubeadm to use client-go wherever possible

**Which issue this PR fixes**: fixes #https://github.com/kubernetes/kubeadm/issues/52

**Special notes for your reviewer**: /cc @luxas @pires 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by mikedanese** [comment]
/approve

, lgtm when tests are green

**Comment by dmmcquay** [comment]
@k8s-bot bazel test this

**Comment by dmmcquay** [comment]
@mikedanese Is there something I am supposed to update in some file? I don't understand why the tests keep failing.

**Comment by pires** [comment]
I believe `kubeadm.k8s.io` is not installed as an API.

**Comment by luxas** [comment]
@lavalamp @sttts @deads2k PTAL
I guess it's the `testapi` that's haunting us, can you please look how we can workaround not having an apigroup registered in client-go, but still use client-go. I've heard folks saying that specific test is a little bit "stupid" in that sense, is there anything we can easily do to make it smarter?

**Comment by dmmcquay** [comment]
@mikedanese tests are green.

**Comment by luxas** [comment]
/approve

I'm applying @mikedanese's lgtm

File: pkg/api/testapi/testapi.go [hunk 1]
```diff
@@ -268,15 +266,6 @@ func init() {
 			externalTypes:        api.Scheme.KnownTypes(externalGroupVersion),
 		}
 	}
-	if _, ok := Groups[kubeadm.GroupName]; !ok {
```
Comment (mikedanese) [inline pkg/api/testapi/testapi.go]: Why is the necessary? Are round trip tests still running?

**Comment by luxas** [comment]
It seems like testapi doesn't have any OWNERS, and it's very awkward to ping pkg/api owners for this, we're just removing kubeadm from the testapi. @deads2k suggested and approved this already on Slack as well

File: pkg/api/testapi/testapi.go (hunk 1 above)
Comment (luxas) [inline pkg/api/testapi/testapi.go]: We couldn't get it working any other way, and I don't think we want to add the kubeadm API group to client-go. See the thread on Slack, please.

**Comment by mikedanese** [comment]
I'm nacking this change until we make sure that test coverage is preserved. Also please squash fixup commits.

**Comment by luxas** [comment]
@mikedanese I asked for a better solution above, but didn't get an answer.
What are you proposing to do instead of removing it?
Should we consider moving the testapi somehow to under `cmd/kubeadm`?

**Comment by mikedanese** [comment]
If moving onto client-go requires turning off all that testing, that's basically a non-starter IMO. I don't know what the solution is since I haven't looked at the specifics. I would create a github issue and ping sig-api-machinary. What does this change actually buy us? I understand the code health benefit but the loss of api group testing is *huge*. Let's do some cost/benefit analysis here, and find the right solution.

**Comment by pires** [comment]
@mikedanese I totally agree and I'd rather stop moving to client-go if and implement our own mock for clientset, if it doesn't exist already somewhere else for the generated clientset. Losing API tests is a no-go for me!

**Comment by deads2k** [comment]
> If moving onto client-go requires turning off all that testing, that's basically a non-starter IMO. I don't know what the answer is since I haven't looked at the specifics. I would create a github issue and ping sig-api-machinary. What does this change actually buy us? I understand the code health benefit but the loss of api group testing is huge. Let's do some cost/benefit analysis here, and find the right solution.

`testapi` doesn't actually test any APIs.  It doesn't even control which APIs are used for testing in the vast majority of the code.  If you aren't using it directly for locating your single version (very unlikely), then remove it has no effect.

File: cmd/kubeadm/app/cmd/version.go [hunk 2]
```diff
@@ -22,8 +22,8 @@ import (
 
 	"github.com/spf13/cobra"
 
+	"k8s.io/client-go/pkg/version"
```
Comment (deads2k) [inline cmd/kubeadm/app/cmd/version.go]: This is an ok transition for now, but it really seems like `kubeadm`'s version should be controlled by kubeadm, not whatever level of the client you vendored.

File: pkg/api/testapi/testapi.go (hunk 1 above)
Comment (deads2k) [inline pkg/api/testapi/testapi.go]: You shouldn't even exist in the same scheme, so the existing round trip tests shouldn't've have picked you up.  In addition, you want a roundtripping test regardless of what the current API registry has registered for versions, so if the fuzzer somehow got wired to `testapi` that was likely a mistake.

Comment (deads2k) [inline pkg/api/testapi/testapi.go]: And yet, that's where we arrived.  How frustrating.  Alright, I'll unwind the fuzzing tests either today or early next week.

**Comment by deads2k** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/41201 to fix the fuzzing.  It gives you enough to create your own scheme and simply call a "fuzz me"

File: cmd/kubeadm/app/cmd/version.go (hunk 2 above)
Comment (mikedanese) [inline cmd/kubeadm/app/cmd/version.go]: We should continue to use k8s.io/kubernetes/pkg/version while we are in the main repo.

**Comment by mikedanese** [comment]
Have we confirmed that apigroup tests still run after this change?

**Comment by dmmcquay** [comment]
@mikedanese I added them back in, but I'm waiting to see if I did it correctly.

**Comment by mikedanese** [comment]
/lgtm

could this use a squash?

**Comment by dmmcquay** [comment]
@mikedanese so I think I messed up by adding the tests back into testapi.go since @deads2k added this in his recent PR: https://github.com/kubernetes/kubernetes/pull/41201/files#diff-60bc03cc4906753953a1a4058faeb37fR164

Am I correct in thinking that? Can I remove what I had done before in testapi.go?

**Comment by dmmcquay** [comment]
@k8s-bot cri node e2e test this

**Comment by dmmcquay** [comment]
@mikedanese had to rebase, lost your lgtm label.

**Comment by mikedanese** [comment]
I think we need to do a bit more to enable testing if you are going to remove kubeadm from testapi. Let me take a look.

**Comment by deads2k** [comment]
> I think we need to do a bit more to enable testing if you are going to remove kubeadm from testapi. Let me take a look.

I saw benchmarking (don't think we care), copying (there's a copy test in the fuzzer), protobuf (kubeadm has no proto), swagger (you don't have a server), ListTypes reflective check (this should be rewritten to operate against a normal scheme, but I wouldn't block this on it).  Nothing else stood out.

**Comment by dmmcquay** [comment]
@mikedanese Can we push this through now since #41352 was merged?

**Comment by luxas** [comment]
LGTM

**Comment by mikedanese** [comment]
Can you rebase and squash?

**Comment by luxas** [comment]
Yea it makes sense to have this as an atomic change

**Comment by dmmcquay** [comment]
@mikedanese rebased and squashed.

**Pull request merged**